File: dynabuffers-java/src/main/kotlin/dynabuffers/Registry.kt
package dynabuffers

import dynabuffers.api.IAnnotation
import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import dynabuffers.ast.ClassType
import dynabuffers.ast.EnumType
import dynabuffers.ast.UnionType
import dynabuffers.ast.annotation.*
import dynabuffers.ast.structural.Annotation
import dynabuffers.exception.DynabuffersException

class Registry(
    private val tree: List<IType>,
    private val listeners: ArrayList<(String) -> Unit>
) : IRegistry {

    override fun resolveAnnotation(annotation: Annotation): IAnnotation {
        when (annotation.options.name) {
            "NotBlank" -> return NotBlank(annotation.options.values)
            "MinLength" -> return MinLength(annotation.options.values)
            "MaxLength" -> return MaxLength(annotation.options.values)
            "GreaterThan" -> return GreaterThan(annotation.options.values)
            "GreaterEquals" -> return GreaterEquals(annotation.options.values)
            "LowerThan" -> return LowerThan(annotation.options.values)
            "LowerEquals" -> return LowerEquals(annotation.options.values)
            else -> throw DynabuffersException("unknown annotation ${annotation.options.name}")
        }
    }

    override fun addNotification(notification: String) {
        listeners.forEach { it(notification) }
    }

    override fun createForSchema(schema: List<IType>): IRegistry {
        return Registry(schema, listeners)
    }

    override fun resolve(name: String): ISerializable {
        return tree.filterIsInstance<ISerializable>().find {
            if (it is ClassType && it.options.name == name) return@find true
            if (it is EnumType && it.options.name == name) return@find true
            if (it is UnionType && it.options.name == name) return@find true

            return@find false
        } ?: throw DynabuffersException("invalid reference $name")
    }
}

File: dynabuffers-python/dynabuffers/Registry.py
from dynabuffers.api.ISerializable import ISerializable
from dynabuffers.ast.ClassType import ClassType
from dynabuffers.ast.EnumType import EnumType
from dynabuffers.ast.UnionType import UnionType
from dynabuffers.ast.annotation.GreaterEquals import GreaterEquals
from dynabuffers.ast.annotation.GreaterThan import GreaterThan
from dynabuffers.ast.annotation.LowerEquals import LowerEquals
from dynabuffers.ast.annotation.LowerThan import LowerThan
from dynabuffers.ast.annotation.MaxLength import MaxLength
from dynabuffers.ast.annotation.MinLength import MinLength
from dynabuffers.ast.annotation.NotBlank import NotBlank


class Registry(object):

    def __init__(self, tree: [ISerializable], listeners):
        self.tree = tree
        self.listeners = listeners

    def resolveAnnotation(self, name, args):
        if name == "GreaterThan":
            return GreaterThan(args)
        if name == "GreaterEquals":
            return GreaterEquals(args)
        if name == "LowerThan":
            return LowerThan(args)
        if name == "LowerEquals":
            return LowerEquals(args)
        if name == "MaxLength":
            return MaxLength(args)
        if name == "MinLength":
            return MinLength(args)
        if name == "NotBlank":
            return NotBlank(args)
        raise ValueError("unknown annotation " + str(name))

    def createForSchema(self, schema: [ISerializable]):
        return Registry(schema, self.listeners)

    def resolve(self, name: str):
        for element in self.tree:
            if isinstance(element, ClassType) and element.options.name == name:
                return element
            if isinstance(element, EnumType) and element.options.name == name:
                return element
            if isinstance(element, UnionType) and element.options.name == name:
                return element

    def addNotification(self, notification: str):
        for listener in self.listeners:
            listener(notification)
